Test /permissive- with Visual Studio 2017

Recent versions of Visual C++ 2017 have a new flag to enable a better conformance to the C++ norm: `/permissive-`.

https://blogs.msdn.microsoft.com/vcblog/2016/11/16/permissive-switch/

I think we should try this flag, in on of our VC++2017.

IMO, that is not a priority task, but it could be done during the reorganization of our Windows testsuite.

`Nef_polyhedron_3` can't be #included with this flag on.  Simply create an empty visual studio 2017 project with the conformance mode (`/permissive-`) turned on (can be found in the C++/Language project setting), simple compile a single translation unit with `#include <CGAL/Nef_polyhedron_3.h>` and it fails.

Link for information about the flag is [here](https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance).

Here's a sample of the errors:

```
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\internal\static_filters\angle_3.h(156): error C2589: '(': illegal token on right side of '::'
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\internal\static_filters\angle_3.h(164): note: see reference to class template instantiation 'CGAL::internal::Static_filters_predicates::Angle_3<K_base>' being compiled
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\internal\static_filters\do_intersect_3.h(321): error C2589: '(': illegal token on right side of '::'
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\internal\static_filters\do_intersect_3.h(327): note: see reference to class template instantiation 'CGAL::internal::Static_filters_predicates::Do_intersect_3<K_base,SFK>' being compiled
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_polyhedron_3.h(224): error C3203: 'SNC_structure': unspecialized class template can't be used as a template argument for template parameter 'Map', expected a real type
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_polyhedron_3.h(2021): note: see reference to class template instantiation 'CGAL::Nef_polyhedron_3<K,I,M>' being compiled
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_polyhedron_3.h(1066): error C3203: 'SNC_structure': unspecialized class template can't be used as a template argument for template parameter 'R', expected a real type
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_polyhedron_3.h(407): error C3203: 'SNC_structure': unspecialized class template can't be used as a template argument for template parameter 'Map', expected a real type
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_polyhedron_3.h(407): note: This diagnostic occurred in the compiler generated function 'CGAL::Nef_polyhedron_3<K,I,M>::Nef_polyhedron_3(II,II,const CGAL::Vector_3<KN> &,bool)'
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_const_decorator.h(57): error C2955: 'CGAL::SNC_structure': use of class template requires template argument list
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_sphere_map.h(44): note: see declaration of 'CGAL::SNC_structure'
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_decorator.h(53): note: see reference to class template instantiation 'CGAL::SNC_const_decorator<Map>' being compiled
1>        with
1>        [
1>            Map=int
1>        ]
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_point_locator.h(69): note: see reference to class template instantiation 'CGAL::SNC_decorator<int>' being compiled
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_point_locator.h(154): note: see reference to class template instantiation 'CGAL::SNC_point_locator<SNC_decorator>' being compiled
1>        with
1>        [
1>            SNC_decorator=CGAL::SNC_decorator<int>
1>        ]
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_polyhedron_3.h(409): note: see reference to class template instantiation 'CGAL::SNC_point_locator_by_spatial_subdivision<CGAL::SNC_decorator<int>>' being compiled
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_const_decorator.h(62): error C2955: 'CGAL::SNC_structure': use of class template requires template argument list
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_sphere_map.h(44): note: see declaration of 'CGAL::SNC_structure'
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_const_decorator.h(67): error C2955: 'CGAL::SNC_structure': use of class template requires template argument list
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_sphere_map.h(44): note: see declaration of 'CGAL::SNC_structure'
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_const_decorator.h(68): error C2955: 'CGAL::SNC_structure': use of class template requires template argument list
1>c:\users\jessepepper\source\repos\vcpkg\installed\x64-windows\include\cgal\nef_3\snc_sphere_map.h(44): note: see declaration of 'CGAL::SNC_structure'
```

I've found some issues compiling with /permissive-
`` c:\dev\cgal-4.11\include\cgal\sweep_line_2\arr_overlay_traits_2.h(410): error C2760: syntax error: unexpected token 'identifier', expected ';' ``
``c:\dev\cgal-4.11\include\cgal\aabb_traits.h(290): error C3861: 'longest_axis': identifier not found``

Those are the start.

Tested with CGAL 4.11 and 4.12beta1/2  and boost 1.65_1 and 1.67_beta All fail with /permissive- but are fine without the /permissive- flag

Hi,

Ive tried the latest 4.12 release and I still get the same exact issue. test code as aobve

> c:\dev\cgal-4.12\include\cgal\aabb_traits.h(305): error C3861: 'longest_axis': identifier not found

compile command

`2>    D:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\VC\Tools\MSVC\14.14.26412\bin\HostX86\x64\CL.exe /c /IC:\dev\QtPCB_msvc\Sandbox /IC:\dev\QtPCB\build\CMake\Sandbox /IC:\dev\QtPCB_msvc\Sandbox\Sandbox_autogen\include_Debug /IC:\dev\QtPCB\build\CMake\Sandbox\..\..\..\include /IC:\local\boost_1_67_0 /I"C:\dev\CGAL-4.12\auxiliary\gmp\include" /I"C:\dev\CGAL-4.12\include" /I"C:\dev\QtPCB\build\CMake\Sandbox\..\..\..\external\explicit-9.1.2018\include" /I"C:\dev\QtPCB\build\CMake\Sandbox\..\..\..\external\multidim-2.2.2018" /ID:\Qt\5.11.0\msvc2017_64\include /ID:\Qt\5.11.0\msvc2017_64\include\QtCore /I"D:\Qt\5.11.0\msvc2017_64\.\mkspecs\win32-msvc" /ID:\Qt\5.11.0\msvc2017_64\include\QtOpenGL /ID:\Qt\5.11.0\msvc2017_64\include\QtWidgets /ID:\Qt\5.11.0\msvc2017_64\include\QtGui /ID:\Qt\5.11.0\msvc2017_64\include\QtANGLE /ID:\Qt\5.11.0\msvc2017_64\include\QtSvg /Zi /nologo /W3 /WX- /diagnostics:classic /MP4 /Od /Ob0 /D WIN32 /D _WINDOWS /D _SCL_SECURE_NO_WARNINGS /D _SCL_SECURE_NO_DEPRECATE /D QT_CORE_LIB /D QT_USE_QSTRINGBUILDER /D QT_NO_KEYWORDS /D CMAKE_PROJECT_NAME=Sandbox /D _CRT_SECURE_NO_WARNINGS /D BOOST_ALL_DYN_LINK /D BOOST_TEST_DYN_LINK /D CGAL_USE_MPFR /D CGAL_USE_GMP /D CGAL_USE_CORE=1 /D BOOST_ALL_DYN_LINK=1 /D QT_OPENGL_LIB /D QT_WIDGETS_LIB /D QT_GUI_LIB /D QT_SVG_LIB /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:strict /fp:except- /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++latest /Fo"Sandbox.dir\Debug\\" /Fd"Sandbox.dir\Debug\vc141.pdb" /Gd /TP /wd4503 /FC /errorReport:prompt  -Zc:throwingNew /bigobj /permissive- C:\dev\QtPCB\source\Sandbox\main.cpp
`

I'm using a shared lib build of CGAL, with all 4 libs built, no changes made in the cmake config.

--- a/AABB_tree/include/CGAL/AABB_traits.h
+++ b/AABB_tree/include/CGAL/AABB_traits.h
@@ -306,13 +306,13 @@ public:
         switch(Traits::longest_axis(bbox))
         {
         case AT::CGAL_AXIS_X: // sort along x
-          std::nth_element(first, middle, beyond, boost::bind(less_x,_1,_2,m_traits));
+          std::nth_element(first, middle, beyond, boost::bind(Traits::less_x,_1,_2,m_traits));
           break;
         case AT::CGAL_AXIS_Y: // sort along y
-          std::nth_element(first, middle, beyond, boost::bind(less_y,_1,_2,m_traits));
+          std::nth_element(first, middle, beyond, boost::bind(Traits::less_y,_1,_2,m_traits));
           break;
         case AT::CGAL_AXIS_Z: // sort along z
-          std::nth_element(first, middle, beyond, boost::bind(less_z,_1,_2,m_traits));
+          std::nth_element(first, middle, beyond, boost::bind(Traits::less_z,_1,_2,m_traits));
           break;
         default:
           CGAL_error();

--- a/AABB_tree/include/CGAL/AABB_traits.h
+++ b/AABB_tree/include/CGAL/AABB_traits.h
@@ -291,7 +291,8 @@ public:
    */
   class Sort_primitives
   {
-  const AABB_traits<GeomTraits,AABBPrimitive,BboxMap>& m_traits;
+    typedef AABB_traits<GeomTraits,AABBPrimitive,BboxMap> Traits;
+    const Traits& m_traits;
   public:
     Sort_primitives(const AABB_traits<GeomTraits,AABBPrimitive,BboxMap>& traits)
       : m_traits(traits) {}
@@ -302,7 +303,7 @@ public:
                     const typename AT::Bounding_box& bbox) const
       {
         PrimitiveIterator middle = first + (beyond - first)/2;
-        switch(longest_axis(bbox))
+        switch(Traits::longest_axis(bbox))
         {
         case AT::CGAL_AXIS_X: // sort along x
           std::nth_element(first, middle, beyond, boost::bind(less_x,_1,_2,m_traits));

